### ISSUE TITLE
(enh) Allow modmaps to be concurrent? 

### ISSUE THREAD

**Issue by RedBearAK** [joshgoebel/keyszer#86]
Implementation Priority: Low

Problem: modmaps have to operate in mutually exclusive conditions, unlike keymaps which are allowed to operate concurrently. 

Use cases like Kinto need to have nearly duplicate modmaps just to make a slight change to basic modifier remaps for specific applications. In Kinto's case, GUI apps get the Super/Win key remapped onto the physical Ctrl keys, while terminal apps get the physical Ctrl keys mapped to Left_Ctrl and no Super/Win key is available. 

If modmaps could be concurrent, as little as a single key remap could be overridden by a secondary modmap, similar to how a keymap that comes before another keymap can override one or more keys or combos in the existing keymap. A short modmap like this would be able to change the Ctrl keys back to Left_Ctrl for terminals in Kinto. Currently the non-terminal and terminal modmaps for Kinto contain the same number of lines just to make a minor change. (This is oversimplified since there are lines for four different keyboard types in Kinto's modmaps, but the principle remains.)

```python
define_conditional_modmap(re.compile(termStr, re.IGNORECASE), {
    Key.LEFT_CTRL: Key.LEFT_CTRL,   # WinMac
    Key.RIGHT_CTRL: Key.LEFT_CTRL,  # WinMac - Multi-language (Remove)
})
```

A secondary example is my attempt to use a modmap to just remap the keypad keys while `numlock_on` is False. This did not work as expected because the modmap focused only on the keypad would clobber the global modmap when it was active. Or vice versa, depending on which one came first in the config file. 

Question is if there would be any unintended difficulties with conflicts if modmaps are concurrent. Technically there are already conflicts with multiple modmaps, so I think it would be an improvement to just make modmaps work like keymaps.

**Comment by joshgoebel** [comment]
Would a single global modmap + many conditional modmaps help?

**Comment by RedBearAK** [comment]
> Would a single global modmap + many conditional modmaps help/be simpler?

I don't really know what you mean by that. The main modmap in Kinto's config is commented as a "global modmap" but it's still using a conditional definition just like the terminal modmap. So I'm not sure I even understand what a "global modmap" is supposed to be, separate from conditionals.

**Comment by joshgoebel** [comment]
Right now (for Keyszer) there is a default modmap (it's empty if not specified) and then conditional modmaps.

The idea is/was that MOST things should go into the default modmap with per-app conditionals for edge cases.  I guess Kinto defines two big "edge case", terminals and not-terminals.  Really the primary (non-terminal) case should be the default modmap.

**Comment by joshgoebel** [comment]
Right now:

- find all modmaps that match conditions
- first modmap that matches is active
- if no matches default is active

Proposal:

- find all modmaps that include keystroke
- filter by modmaps that ALSO match conditions
- first modmap is active
- if no matches default is active

**Comment by RedBearAK** [comment]
> So if the keypad is unique I think it would mean you'd only need two modmaps? (one for on and one for off?)

Current situation with making both the GTK3 numpad nav keys fix and the "Mac Numpad" feature work is that I have six mutually exclusive conditional modmaps. One each for Numlock ON, OFF, and "ignore Numlock and always be numpad". After the fix for the attribute error, which for some reason only came up when I tried to do the terminal modmaps, this is all working as intended. 

I've moved most of the functionality of the "Mac Numpad" feature into the two modmaps that engage when the `_mac_numpad` variable is True, since that more closely creates how things work on a real Mac keyboard. It was all in a keymap at first. 

> The idea is/was that MOST things should go into the default modmap with per-app conditionals for edge cases.

Are we just talking here about the difference between what `define_modmap` was supposed to do, and what `define_conditional_modmap` is supposed to do? And then in the newer `modmap` it's just "global" if there are no conditions attached?

**Comment by RedBearAK** [comment]
This is what the whole set of modmaps looks like right now. 

[snippet_new_kinto_modmaps.py.txt](https://github.com/joshgoebel/keyszer/files/9419665/snippet_new_kinto_modmaps.py.txt)

**Comment by joshgoebel** [comment]
> Numlock ON, OFF, and "ignore Numlock and always be numpad"

Still sounds like two states/behaviors to me, numpad ON or OFF - you're just explaining HOW...

> And then in the newer modmap it's just "global" if there are no conditions attached?

Yes.  Only a single default modmap is allowed.

**Comment by joshgoebel** [comment]
> This is what the whole set of (six) modmaps looks like right now.

Yeah, this is ugly and annoying. :-)  I feel like you should be able to just have:

```py
modmap("numpad ON (for any reason)",  {
}, when = ... )

modmap("numpad OFF (for any reason)",  {
}, when = ...)
```

And I think one of them could possibly even be empty?  That seems to be what you're doing with Kinto...

**Comment by joshgoebel** [comment]
Did you read my proposal: https://github.com/joshgoebel/keyszer/issues/86#issuecomment-1225631348

I think that would perhaps button this up and be a small change.

**Comment by RedBearAK** [comment]
> Still sounds like two states/behaviors to me, numpad ON or OFF - you're just explaining HOW...

I tried with just two, but it didn't seem to work out well because of the whole "mutually exclusive" way the modmaps seem to operate at the moment. 

> Did you read my proposal

I looked it over, but the only context I have for understanding it is if the proposal is more like how keymaps already work. 

> I think that would perhaps button this up and be a small change.

If you think the change would be minimal and you're willing to implement it, I will do my best to test it with a simpler group of "overlapping" modmaps. I have no doubt that they could and should be a lot more straightforward. 

What I would expect to be able to use to do everything I'm trying to do in Kinto would be something like...

```python
modmap("Mac Numpad feature enabled",  {
# just numpad mappings (numeric)
}, when = _mac_numpad is True)

modmap("Numlock OFF",  {
# just numpad mappings (nav key functions)
}, when = numlock_on is False)

modmap("Terminals overrides",  {
# terminals Ctrl key override of general modmap
}, when = in terminals)

modmap("Global modmap",  {
# general mappings
})
```

With the only big one being the "global" modmap.

**Comment by joshgoebel** [comment]
> If you think the change would be minimal 

It's literally a single line. :-)  I'm just trying to think of any downsides... right now this "new way" seems a lot nicer/simpler for organization modmaps.

**Comment by RedBearAK** [comment]
> It's literally a single line.

"Ah, but the hard part is knowing where to put the line," said the mechanic. "That'll be $1,500."

> I'm just trying to think of any downsides

Should still be easy to create mutually-exclusive conditions, if that's the way some users are used to having it work, right? 

There's really not much distinction between modmaps and keymaps in AHK. You either have... 

This key _is_ this key (modmap), or 
This key/combo _sends_ this key/combo (keymap)

And then you wrap them in whatever directives you want to create the conditions.

**Comment by joshgoebel** [comment]
@RedBearAK Please test with your numlock stuff.  There is a passing test in the test suite as well.

**Comment by RedBearAK** [comment]
> @RedBearAK Please test with your numlock stuff. There is a passing test in the test suite as well.

Had to disable the whole Option key scheme in my config since I've modified it so much to work with the immediate trigger, which hasn't been merged with main yet. 

Other than that, moving onto the concurrency branch everything seems to work as expected in the six different conditions. 

Haven't yet tried modifying the modmaps to shrink all but the main "global" modmap. That will take a while.

**Comment by RedBearAK** [comment]
@joshgoebel 

300 lines down to about 110. Seems to be working as expected from my time working with overlapping keymaps. 

The other Kinto keyboard types blocks have some lines that appear to rely on the original behavior of the main modmap getting completely disabled when the terminal modmap is active, but that should be easily fixable by just explicitly defining the overriding modmaps for the left Ctrl key for all keyboard types, as it already is for typical "PC" keyboards. 

Other than that, all preliminary testing of the rearranged and trimmed modmaps shows good results. After I remembered to make the main one truly "global" (i.e., no conditions at all) so that it would do the main remaps in terminals as well as everywhere else. 

[snippet_new_kinto_modmaps_concurrent.py.txt](https://github.com/joshgoebel/keyszer/files/9438824/snippet_new_kinto_modmaps_concurrent.py.txt)

Oh, and I can finally remove **_all_** the remaining "insane" numbers of remaps for fixing the numpad nav keys in GTK3 apps. Now it can just be a modmap, active only in the appropriate "Numlock OFF" circumstances. That, plus the media arrow keys fix turning into a modmap, removes a few hundred lines from my config file.

**Comment by joshgoebel** [comment]
You shouldn't need to modify any of the existing modmaps. You should only need to add two new maps for both keypad states.

**Comment by RedBearAK** [comment]
> You shouldn't need to modify any of the existing modmaps. You should only need to add two new maps for both keypad states.

I figured part of the goal here was to remove unnecessary lines. I get what you're saying, I could leave the original modmaps in place and keep the mutually-exclusive conditions. But I don't really see the point of doing it that way. Unless you want to reduce the number of modmaps active at the same time? 

> That's what I was suggesting you test if it works with just two numpad maps as it should.

That's pretty much what I did, if you ignore the modified terminals modmap. And it does work as expected. Including how modmaps defined earlier will override later modmaps with the same remaps when they become active. Since it modified the terminals modmap that particular aspect has been tested twice.

**Comment by joshgoebel** [comment]
> The other Kinto keyboard types blocks have some lines that appear to rely on the original behavior of the main modmap getting completely disabled when the terminal modmap is active, 

Yeah. The new behavior might require some small tweaks,  but really if a thing isn't truly global it probably shouldn't be in the default modmap in the first place. 

So the Ctrl behavior likely does not belong there since it varies depending on app. 

So you think this is good to merge?

**Comment by RedBearAK** [comment]
> So you think this is good to merge?

I can say that I don't appear to be having any issues with it. 

If we can get the immediate trigger also merged to main I'll be almost on Easy Street, other than still needing a way to pass through input unmodified from a nested construct. 

> So the Ctrl behavior likely does not belong there since it varies depending on app.

🤔 Hmmm. I mean, it is dependent on being in a class of apps (i.e., terminals), but I think it still needs to be a modmap. So I'm not really sure what you mean by this. 

I'll show you the modified terminal modmap when I get finished adapting it. Should be significantly smaller than it was, and already has the "media arrows fix" removed so that it isn't redundant with the global block.

**Comment by RedBearAK** [comment]
This should take care of the terminal block behaving for all keyboard types (and multi-language) like it was originally intended, with the main modmap remaining truly global with no conditions. 

[snippet_new_kinto_modmaps_concurrent.py.txt](https://github.com/joshgoebel/keyszer/files/9438909/snippet_new_kinto_modmaps_concurrent.py.txt)

**Comment by joshgoebel** [comment]
> So I'm not really sure what you mean by this.

I was imagining:

- default (always)
- case X
- case Y

And if something wasn't always true you'd avoid putting it in default - but really this is a users choice...

**Comment by joshgoebel** [comment]
> other than still needing a way to pass through input unmodified from a nested construct.

Can you find your notes on this and create a new issue just for this feature?

**Comment by RedBearAK** [comment]
> Can you find your notes on this and create a new issue just for this feature?

@joshgoebel 

I think it's already encapsulated in issue #63, which was branched off of issue #61 (closed). Issue #65 also came from that, which should be resolved with the merging of the immediate trigger.

**Comment by RedBearAK** [comment]
@joshgoebel 

Just confirming, I switched back to the main branch, did a pull and both the modified modmaps and the re-enabled Option key scheme with entry triggers are working together. It's great having the basic visual feedback for the dead keys accents. 

I switched over to using a small function to set the "ac_Chr" variable for the dead keys and it works like a charm. So if you want to go ahead and block "invalid" commands like the way I was assigning the variable value inside the macro, that shouldn't cause me any trouble.
